### ISSUE TITLE
GRIB-reader generating non-UDUNITS units strings

### ISSUE THREAD

**Issue by BobSimons** [Unidata/thredds#454]
[This is related to issue #449, where the units generated are degrees.true (which is not a valid UDUNITS string).]

When netcdf-java 4.6.4 reads an older version of multi_1.glo_30m.all.grb2 from 
ftp://polar.ncep.noaa.gov/pub/waves/develop/multi_1.latest_run/multi_1.glo_30m.all.grb2
(if the version at the ftp site doesn't show the problem, let me know and I can send you my copy),
there are two small problems with the metadata:

1) For varName=ordered_sequence_of_data, units="seq".
I could be wrong, but I don't think "seq" is a valid UDUNITS units string.
Can you please change that to "count", which is a better (valid and reasonably
appropriate) units string, or whatever valid units string you think is better?

2) For varName=Direction_of_swell_waves_ordered_sequence_of_data, units="degree.true".
That is not a valid UDUNITS units string.
Please change that to "degree_true", which is the valid and appropriate units string.
[This was just reported by someone else as issue #449, where the units generated are degrees.true (not a valid UDUNITS string).]

My suggestion related to this bug report and issue #449: Maybe it would be good to scan the tables you use when decoding GRIB files for units strings that are not valid UDUNITS strings.

Thanks.

**Comment by dopplershift** [comment]
The file you linked to disappeared...will it come back?

**Comment by BobSimons** [comment]
Well, it's probably better if we work from exactly the same file anyway.
I think/hope you can download the version of the file that I test with from
https://drive.google.com/open?id=0B2zEESjdXUmwLThlVTRUUlFnSUU
Let me know when you successfully have it so I can delete that copy.
Or if this doesn't work, let me know and send me your more direct email
address so I can do this a different way.


On Tue, Mar 1, 2016 at 12:38 PM, Ryan May <notifications@github.com> wrote:

> The file you linked to disappeared...will it come back?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Unidata/thredds/issues/454#issuecomment-190890779>.
>



-- 
Sincerely,

Bob Simons
IT Specialist
Environmental Research Division
NOAA Southwest Fisheries Science Center
99 Pacific St., Suite 255A      (New!)
Monterey, CA 93940               (New!)
Phone: (831)333-9878            (New!)
Fax:   (831)648-8440
Email: bob.simons@noaa.gov

The contents of this message are mine personally and
do not necessarily reflect any position of the
Government or the National Oceanic and Atmospheric Administration.
<>< <>< <>< <>< <>< <>< <>< <>< <><

**Comment by dopplershift** [comment]
@BobSimons I have it, thanks.

@JohnLCaron We manually set the units to "seq" at NcepLocalTables.java:338:
```java
        return new GribLevelType(code, "seq", null, true);   // eg see NCEP World Watch datasets
```
Would "count" make sense? To me it's no worse, but I don't have any idea if "seq" was specifically chosen.

**Comment by JohnLCaron** [comment]
"seq" is not a valid udunit. but then, neither is "count".

NCEP has:
    codeMap.put("4.5.241", "Ordered Sequence of Data");

from

http://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_table4-5.shtml

the correct udunit would be blank (dimensionless), but thats not very helpful.

back to why dimensional units, aka udunits, is not the right choice for the units attribute.

**Comment by dopplershift** [comment]
Count's not udunits compatible? Then why is it in the database (from [here](https://www.unidata.ucar.edu/software/udunits/udunits-2.2.20/udunits/udunits2-common.xml)):
```xml
<unit>
  <def>1</def>
  <aliases>
    <name>
      <singular>count</singular>
    </name>
  </aliases>
  <definition/>
</unit>
```

**Comment by BobSimons** [comment]
"count" was a valid UDUNITS-1 unit.
I don't have a way to test UDUNITS-2, but I see
https://www.unidata.ucar.edu/support/help/MailArchives/udunits/msg00766.html
and so hope that "count" is again valid in UDUNITS-2.
"count" is/was very useful.

On Tue, Mar 1, 2016 at 3:26 PM, John Caron <notifications@github.com> wrote:

> "seq" is not a valid udunit. but then, neither is "count".
>
> NCEP has:
> codeMap.put("4.5.241", "Ordered Sequence of Data");
>
> from
>
> http://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_table4-5.shtml
>
> the correct udunit would be blank (dimensionless), but thats not very
> helpful.
>
> back to why dimensional units, aka udunits, is not the right choice for
> the units attribute.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Unidata/thredds/issues/454#issuecomment-190961741>.
>



-- 
Sincerely,

Bob Simons
IT Specialist
Environmental Research Division
NOAA Southwest Fisheries Science Center
99 Pacific St., Suite 255A      (New!)
Monterey, CA 93940               (New!)
Phone: (831)333-9878            (New!)
Fax:   (831)648-8440
Email: bob.simons@noaa.gov

The contents of this message are mine personally and
do not necessarily reflect any position of the
Government or the National Oceanic and Atmospheric Administration.
<>< <>< <>< <>< <>< <>< <>< <>< <><

**Comment by JohnLCaron** [comment]
you are right, sorry i was wrong (sheepish grin)

**Comment by JohnLCaron** [comment]
i guess "count" then is better than "seq". both dont say much about what
the heck it is.

and, oh, btw, its not a vertical coordinate at all, more akin to wave
number i think. oh well.

**Comment by dopplershift** [comment]
I agree, neither helps me understand--but I default to "a" standard rather than ad hoc. I'll whip up a PR for this and "degree.true".

**Comment by BobSimons** [comment]
I hope you will scan for other non-UDUNITS strings while you are poking
around those tables. Please don't make me report them one by one over the
coming years as I find them.

On Tue, Mar 1, 2016 at 3:52 PM, Ryan May <notifications@github.com> wrote:

> I agree, neither helps me understand--but I default to "a" standard rather
> than ad hoc. I'll whip up a PR for this and "degree.true".
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Unidata/thredds/issues/454#issuecomment-190972515>.
>



-- 
Sincerely,

Bob Simons
IT Specialist
Environmental Research Division
NOAA Southwest Fisheries Science Center
99 Pacific St., Suite 255A      (New!)
Monterey, CA 93940               (New!)
Phone: (831)333-9878            (New!)
Fax:   (831)648-8440
Email: bob.simons@noaa.gov

The contents of this message are mine personally and
do not necessarily reflect any position of the
Government or the National Oceanic and Atmospheric Administration.
<>< <>< <>< <>< <>< <>< <>< <>< <><

**Comment by dopplershift** [comment]
Well the problem is that the tables don't contain any udunits-compatible strings. We're programmatically setting them. In the case of "true.degree", the code to fix non-udunits (turn space into multiplication) was wrong in this case, but is correct 90% the time. The "seq" issue is in our code specifically handling NCEP's "special" codes.

If you have some way you're detecting non-udunits compatible units in the output from netcdf-java, without a human deciding it, I'm all ears.

**Comment by BobSimons** [comment]
90% isn't reassuring.

I don't know your exact situation.
You have the grib info tables. Don't they contain all possible source units
strings?
So can't you, one time, run all the units strings in them through your
rules and look for non-udunits results?  (Either programmatically run them
into the UDUNITS software or, if that's not possible, a human should be
able to scan them reasonably quickly.)
And for each invalid result, can't you then write another rule?
Like I said, I don't know the situation. I'm just guessing. But it seems
doable and worth doing to me.


On Tue, Mar 1, 2016 at 4:27 PM, Ryan May <notifications@github.com> wrote:

> Well the problem is that the tables don't contain any udunits-compatible
> strings. We're programmatically setting them. In the case of "true.degree",
> the code to fix non-udunits (turn space into multiplication) was wrong in
> this case, but is correct 90% the time. The "seq" issue is in our code
> specifically handling NCEP's "special" codes.
>
> If you have some way you're detecting non-udunits compatible units in the
> output from netcdf-java, without a human deciding it, I'm all ears.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Unidata/thredds/issues/454#issuecomment-190986782>.
>



-- 
Sincerely,

Bob Simons
IT Specialist
Environmental Research Division
NOAA Southwest Fisheries Science Center
99 Pacific St., Suite 255A      (New!)
Monterey, CA 93940               (New!)
Phone: (831)333-9878            (New!)
Fax:   (831)648-8440
Email: bob.simons@noaa.gov

The contents of this message are mine personally and
do not necessarily reflect any position of the
Government or the National Oceanic and Atmospheric Administration.
<>< <>< <>< <>< <>< <>< <>< <>< <><

**Comment by JohnLCaron** [comment]
The deeper problem is that udunits is inadequate for expressing some units.

We've been considering adding another attribute to preserve as much info as possible, eg "degree true".

**Comment by dopplershift** [comment]
@BobSimons I'm curious as to the details of your workflow that makes udunit-compatible units so critical-sounding? Are there programs that are failing as a result?

**Comment by BobSimons** [comment]
On Wed, Mar 2, 2016 at 9:17 AM, John Caron <notifications@github.com> wrote:

> The deeper problem is that udunits is inadequate for expressing some units.
>
Like?
Isn't that a reason to add to udunits?

> We've been considering adding another attribute to preserve as much info
> as possible, eg "degree true".
>
Huh? degrees_true is a valid udunit.

—
> Reply to this email directly or view it on GitHub
> <https://github.com/Unidata/thredds/issues/454#issuecomment-191331668>.
>



-- 
Sincerely,

Bob Simons
IT Specialist
Environmental Research Division
NOAA Southwest Fisheries Science Center
99 Pacific St., Suite 255A      (New!)
Monterey, CA 93940               (New!)
Phone: (831)333-9878            (New!)
Fax:   (831)648-8440
Email: bob.simons@noaa.gov

The contents of this message are mine personally and
do not necessarily reflect any position of the
Government or the National Oceanic and Atmospheric Administration.
<>< <>< <>< <>< <>< <>< <>< <>< <><

**Comment by BobSimons** [comment]
Seriously?
I assume you know that CF requires that units strings be valid udunits
units.

1) Aren't standards like CF and udunits made so that software can work with
the (in this case) units without human intervention or that humans can
understand the data via a consistent, defined units system?  Do you want to
go back to everyone writing units however they want? It would be a terrible
mess.
Isn't that reason enough?

2) Isn't the point of standards to follow the standards? Isn't that reason
enough?

3) Aren't CDF and CF entwined? Is there any reason for Unidata software
(netcdf-java) to be generating units that aren't valid udunits (which are
required units for the CF standard) when valid udunits are available?

4) "programs that are failing"? Sort of.  ERDDAP and THREDDS support DAP
services and OGC services. Most of the datasets in ERDDAP and THREDDS
follow CF conventions, hence udunits. I don't know about THREDDS, but when
ERDDAP sends out OGC info, it converts the udunits to ucum units via a
converter that is built into ERDDAP. (Public access via
http://coastwatch.pfeg.noaa.gov/erddap/convert/units.html )
(Isn't that the code you stole from me ;-) without crediting me? Or was it
one of the other converters in ERDDAP?)
So if some units in a dataset aren't udunits, ERDDAP won't be able to
convert them properly to ucum units.  It won't break or throw and
exception. It just passes them through (perhaps as a comment). So the
resulting ucum units won't be correct.

Some people at Unidata have a bizarre attitude about standards, as in "we
own the stack, we can do anything we want". I think that's the wrong
attitude. That is one of the reasons I use Unidata software as little as
possible and one of the reasons I started making ERDDAP.



On Wed, Mar 2, 2016 at 9:26 AM, Ryan May <notifications@github.com> wrote:

> @BobSimons <https://github.com/BobSimons> I'm curious as to the details
> of your workflow that makes udunit-compatible units so critical-sounding?
> Are there programs that are failing as a result?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Unidata/thredds/issues/454#issuecomment-191337927>.
>



-- 
Sincerely,

Bob Simons
IT Specialist
Environmental Research Division
NOAA Southwest Fisheries Science Center
99 Pacific St., Suite 255A      (New!)
Monterey, CA 93940               (New!)
Phone: (831)333-9878            (New!)
Fax:   (831)648-8440
Email: bob.simons@noaa.gov

The contents of this message are mine personally and
do not necessarily reflect any position of the
Government or the National Oceanic and Atmospheric Administration.
<>< <>< <>< <>< <>< <>< <>< <>< <><

**Comment by lesserwhirls** [comment]
Hi Bob,

Please let me try to address some of the concerns you've brought up here.

"I assume you know that CF requires that units strings be valid udunits units."

Yes, CF requires units to be UDUNITs compatible, and we do know this. In fact, @JohnLCaron is an author of the CF docs. Unfortunately, GRIB does not care one bit about CF. We try to keep output from netcdf-java or the TDS as close to CF as possible, but sometimes we just can't do it.

"Do you want to go back to everyone writing units however they want? "

With GRIB, this is exactly the situation we have. We try to do our best with the units we get from GRIB tables, but ultimately it's a failing endeavor.

From a previous comment, you said:

"So can't you, one time, run all the units strings in them through your rules and look for non-udunits results?"

It would be nice to do this one time, but each time tables update, which again is often, custom rules will break. Each time a GRIB table is updated (which happens often), there is a chance that whomever is updating the table, on a whim, will decide that "hey, why do we use degC for temperature units. That looks ugly. I'm going to make it easier to ready by changing it to C". Well, UDUNITs now things that's coulomb. It's valid. How do we, in an automated fashion, catch this? We cannot rely on "variable names" ether, as these are also subject to human tweaking by the GRIB table creator.

Given the use of GRIB around the world, and all of the centers and subcenters "doing their own thing", overwriting the WMO standard tables whenever convenient, not putting the effort needed to version their tables, etc, Unidata cannot dedicate resources to hand maintaining others' GRIB tables (we are talking thousands and thousands of parameters). We essentially have 2 or so FTE on netCDF-Java and the THREDDS Data Server, so being able to get units to conform to UDUNITS 90% of the time is pretty good. While GRIB is an important data format, it's one of many we try to support and already dwarfs the others as to how much time and resources we put into trying to get it right.

Now, if you are running a TDS and you notice an incorrect unit, NcML can be easily applied to the dataset to correct it. This is best done by the data producer, as they know best what is in their dataset. If you are using netCDF-Java in your library, then NcML can be applied as well.

"I don't know about THREDDS, but when ERDDAP sends out OGC info, it converts the udunits to ucum units via a converter that is built into ERDDAP. (Public access via http://coastwatch.pfeg.noaa.gov/erddap/convert/units.html ) (Isn't that the code you stole from me ;-) without crediting me? Or was it one of the other converters in ERDDAP?)"

Accusing someone of stealing your code without letting them know about it first and trying to make things right is not very collegial. If we have code from you but do not attribute it to you, please let us know and we'll fix that. It's not our intent to 'steal' from others without attribution.

We appreciate all of your feedback, but as this thread grew longer and longer, your tone grew more and more hostile. We're all ultimately trying to do the best we can with the resources given, and we're all striving towards the same goal. Let's keep things civil, please.

**Comment by BobSimons** [comment]
Thanks for the information.

Sorry, yes, I was/am upset by your (Unidata's) seeming lack of knowledge of
udunits, your seeming disinterest in supporting udunits, your seeming
unawareness of my units conversion software that you used in your software,
and for asking me why I care that units follow the udunits standard. I do
appreciate that you have limited resources, so do I. I know that grib is a
messy, troublesome project. Thank you for doing it.

I won't pester you any more about automating the checking of units from
grib files. It is probably harder than I envisioned.

Well, here's my calendar and units code in your (by which I mean Unidata's)
code
http://www.unidata.ucar.edu/software/thredds/v4.5/netcdf-java/javadocAll/ucar/nc2/ogc/erddap/util/package-frame.html
I'm fine with you using it.
You struck a nerve because: when you first did it, you kept the original
class names but moved them from the com.cohort namespace (what I did before
NOAA) and gov.noaa namespace to your namespace. So it mucked up my use of
subsequent versions of netcdf-java. But eventually you fixed that.
And I didn't/don't think it was very collegial of you to use my software
and not credit me/CoHort Software/NOAA.
Yes, it would be nice if you credited me and NOAA and mentioned how the
software is licensed. I don't see my name, or CoHort Software, or NOAA at
http://www.unidata.ucar.edu/software/netcdf/credits.html
(Or maybe it is somewhere else (other than the original info in the source
code) and I haven't seen it. If so, I apologize.)
I credit you and mention how the software is licensed:
http://coastwatch.pfeg.noaa.gov/erddap/download/setup.html#credits

Sorry. I didn't/don't mean to start a flame war. I want to stay on good
terms with Unidata.




On Wed, Mar 2, 2016 at 11:00 AM, Sean Arms <notifications@github.com> wrote:

> Hi Bob,
>
> Please let me try to address some of the concerns you've brought up here.
>
> "I assume you know that CF requires that units strings be valid udunits
> units."
>
> Yes, CF requires units to be UDUNITs compatible, and we do know this. In
> fact, @JohnLCaron <https://github.com/JohnLCaron> is an author of the CF
> docs. Unfortunately, GRIB does not care one bit about CF. We try to keep
> output from netcdf-java or the TDS as close to CF as possible, but
> sometimes we just can't do it.
>
> "Do you want to go back to everyone writing units however they want? "
>
> With GRIB, this is exactly the situation we have. We try to do our best
> with the units we get from GRIB tables, but ultimately it's a failing
> endeavor.
>
> From a previous comment, you said:
>
> "So can't you, one time, run all the units strings in them through your
> rules and look for non-udunits results?"
>
> It would be nice to do this one time, but each time tables update, which
> again is often, custom rules will break. Each time a GRIB table is updated
> (which happens often), there is a chance that whomever is updating the
> table, on a whim, will decide that "hey, why do we use degC for temperature
> units. That looks ugly. I'm going to make it easier to ready by changing it
> to C". Well, UDUNITs now things that's coulomb. It's valid. How do we, in
> an automated fashion, catch this? We cannot rely on "variable names" ether,
> as these are also subject to human tweaking by the GRIB table creator.
>
> Given the use of GRIB around the world, and all of the centers and
> subcenters "doing their own thing", overwriting the WMO standard tables
> whenever convenient, not putting the effort needed to version their tables,
> etc, Unidata cannot dedicate resources to hand maintaining others' GRIB
> tables (we are talking thousands and thousands of parameters). We
> essentially have 2 or so FTE on netCDF-Java and the THREDDS Data Server, so
> being able to get units to conform to UDUNITS 90% of the time is pretty
> good. While GRIB is an important data format, it's one of many we try to
> support and already dwarfs the others as to how much time and resources we
> put into trying to get it right.
>
> Now, if you are running a TDS and you notice an incorrect unit, NcML can
> be easily applied to the dataset to correct it. This is best done by the
> data producer, as they know best what is in their dataset. If you are using
> netCDF-Java in your library, then NcML can be applied as well.
>
> "I don't know about THREDDS, but when ERDDAP sends out OGC info, it
> converts the udunits to ucum units via a converter that is built into
> ERDDAP. (Public access via
> http://coastwatch.pfeg.noaa.gov/erddap/convert/units.html ) (Isn't that
> the code you stole from me ;-) without crediting me? Or was it one of the
> other converters in ERDDAP?)"
>
> Accusing someone of stealing your code without letting them know about it
> first and trying to make things right is not very collegial. If we have
> code from you but do not attribute it to you, please let us know and we'll
> fix that. It's not our intent to 'steal' from others without attribution.
>
> We appreciate all of your feedback, but as this thread grew longer and
> longer, your tone grew more and more hostile. We're all ultimately trying
> to do the best we can with the resources given, and we're all striving
> towards the same goal. Let's keep things civil, please.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Unidata/thredds/issues/454#issuecomment-191373207>.
>



-- 
Sincerely,

Bob Simons
IT Specialist
Environmental Research Division
NOAA Southwest Fisheries Science Center
99 Pacific St., Suite 255A      (New!)
Monterey, CA 93940               (New!)
Phone: (831)333-9878            (New!)
Fax:   (831)648-8440
Email: bob.simons@noaa.gov

The contents of this message are mine personally and
do not necessarily reflect any position of the
Government or the National Oceanic and Atmospheric Administration.
<>< <>< <>< <>< <>< <>< <>< <>< <><

**Comment by cwardgar** [comment]
Hi Bob,

You're right, I did utilize some of your code in THREDDS. You can find it in our GitHub repo [here](https://github.com/Unidata/thredds/tree/master/waterml/src/main/java/ucar/nc2/ogc/erddap/util). However, I think it's unfair to say that we "stole" it without giving "credit". The NOAA and Cohort licenses have always been present along with the code.

It's true that we could probably give more prominent credit, as you have done on your page. I'll look into that.

**Comment by dopplershift** [comment]
I'll note that this page http://www.unidata.ucar.edu/software/netcdf/credits.html is specific to the netcdf-c (and fortran) library--we don't seem to have one for TDS. It's not a bad idea to add one, but I don't know enough of the history to be able to add anyone else.

**Comment by BobSimons** [comment]
I apologize. It was the wrong word for me to use.

On Wed, Mar 2, 2016 at 11:52 AM, Christian W <notifications@github.com>
wrote:

> Hi Bob,
>
> You're right, I did utilize some of your code in THREDDS. You can find it
> in our GitHub repo here
> <https://github.com/Unidata/thredds/tree/master/waterml/src/main/java/ucar/nc2/ogc/erddap/util>.
> However, I think it's unfair to say that we "stole" it without giving
> "credit". The NOAA and Cohort licenses have always been present along with
> the code.
>
> It's true that we could probably give more prominent credit, as you have
> done on your page. I'll look into that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Unidata/thredds/issues/454#issuecomment-191397750>.
>



-- 
Sincerely,

Bob Simons
IT Specialist
Environmental Research Division
NOAA Southwest Fisheries Science Center
99 Pacific St., Suite 255A      (New!)
Monterey, CA 93940               (New!)
Phone: (831)333-9878            (New!)
Fax:   (831)648-8440
Email: bob.simons@noaa.gov

The contents of this message are mine personally and
do not necessarily reflect any position of the
Government or the National Oceanic and Atmospheric Administration.
<>< <>< <>< <>< <>< <>< <>< <>< <><

**Comment by lesserwhirls** [comment]
Hi Bob,

I don't think you are trying to start a flame war - I think you are very frustrated, and that's certainly not something we desire for any of our users. 

The group here at Unidata has underwent quite a bit of change in the last few years, and with the somewhat recent departure of @JohnLCaron from Unidata, and his knowledge of the users of netCDF-Java, we are playing a bit of catch up. When we ask seemingly insulting questions like "what is your workflow for needing UDUNITS", it's because we as individuals are trying to get a handle on who our community is and how they are using the netCDF-Java library to better understand how we can help. As a team, I'd say we don't know a whole lot about ERDDAP, but we do have users who mention and highly praise it (looking at @rsignell-usgs). I think it would be great to have a Google Hangout or something of the likes with you so that we can see exactly what it is that ERDDAP is doing and how we may help, if you are willing.

Part of the reason we went to Github is to be more transparent about development. We recently took advantage of Github's protected branch feature, and now that every commit we make to master has to be done via pull requests (and must pass part of our test suite), we hope to be able to get user feedback on changes we make. Same thing with Github issues - we hope these generate input from users about possible new developments.

Now, circling around, in terms of doing automated unit checking, I think there are things that we can do. The first step will be to scan through all of the GRIB1 and GRIB2 parameter tables and see what is/is not UDUNITS compatible. it's possible that we can improve our checking substantially in a generic way that can tolerate the changes made to GRIB tables at various centers/subcenters, but there will always be issues mapping what is essentially free text input to a standard (I think we all know the difficulty of even trying to map between standards!).

On the topic of units, but out of curiosity, how do you (ERDDAP) handle observations or model output that samples a change in temperature? The UDUNITS package does not address deltas of temperature at all (long standing issue) - it will convert between various units of temperature, but it is not "delta" aware.

**Comment by dopplershift** [comment]
Fixed with #462. Not sure if there's more discussion to be had here, or if this is the right forum. Closing since at least the software issues is addressed.

**Comment by BobSimons** [comment]
Thank you for your email. Please see my comments below.

On Wed, Mar 2, 2016 at 1:06 PM, Sean Arms <notifications@github.com> wrote:

> Hi Bob,
>
> I don't think you are trying to start a flame war - I think you are very
> frustrated, and that's certainly not something we desire for any of our
> users.
>
> The group here at Unidata has underwent quite a bit of change in the last
> few years, and with the somewhat recent departure of @JohnLCaron
> <https://github.com/JohnLCaron> from Unidata, and his knowledge of the
> users of netCDF-Java, we are playing a bit of catch up. When we ask
> seemingly insulting questions like "what is your workflow for needing
> UDUNITS", it's because we as individuals are trying to get a handle on who
> our community is and how they are using the netCDF-Java library to better
> understand how we can help.
>
Sounds good.

> As a team, I'd say we don't know a whole lot about ERDDAP, but we do have
> users who mention and highly praise it (looking at @rsignell-usgs
> <https://github.com/rsignell-usgs>). I think it would be great to have a
> Google Hangout or something of the likes with you so that we can see
> exactly what it is that ERDDAP is doing and how we may help, if you are
> willing.
>
I am willing.
I talk to Ethan Davis some at conferences.
For my work, you help tremendously by making netcdf-java (including its
ability to read GRIB and BUFR), UDUNITS, and the original ACDD, and by your
support for CF and ACDD.  (I'm probably forgetting something obvious.) They
are obviously fundamental parts of this community and ERDDAP.
10 years ago, I was very critical of the lack of Unidata documentation and
the presence of outdated/incorrect documentation, but that problem has been
almost entirely taken care of. Thank you very much.
Although it seems like we should be able to cooperate on things in common
like WMS, WCS 2.0, etc., I'm not sure how. Those pieces have to fit into
very different puzzles (THREDDS and ERDDAP). But I'm open to suggestions.

> Part of the reason we went to Github is to be more transparent about
> development. We recently took advantage of Github's protected branch
> feature, and now that every commit we make to master has to be done via
> pull requests (and must pass part of our test suite), we hope to be able to
> get user feedback on changes we make. Same thing with Github issues - we
> hope these generate input from users about possible new developments.
>
> Now, circling around, in terms of doing automated unit checking, I think
> there are things that we can do. The first step will be to scan through all
> of the GRIB1 and GRIB2 parameter tables and see what is/is not UDUNITS
> compatible. it's possible that we can improve our checking substantially in
> a generic way that can tolerate the changes made to GRIB tables at various
> centers/subcenters, but there will always be issues mapping what is
> essentially free text input to a standard (I think we all know the
> difficulty of even trying to map between standards!).
>
Sounds great. Thank you.

> On the topic of units, but out of curiosity, how do you (ERDDAP) handle
> observations or model output that samples a change in temperature? The
> UDUNITS package does not address deltas of temperature at all (long
> standing issue) - it will convert between various units of temperature, but
> it is not "delta" aware.
>
I/ERDDAP have no special solution for temperature deltas, e.g., I use
degrees_C for temperatures and (probably erroneously) for temperature
deltas since that is the accepted practice (e.g., for humans describing
both types of data) and omitting units for temperature deltas seems worse.
Unlike UDUNITS, ERDDAP currently doesn't convert from one units scale to
another (e.g., degrees_C to degrees_F) (although people ask for it) so
ERDDAP avoids that problem.
It seems like there could/should be new udunits that identify temperature
deltas (e.g., delta_degree_C?) and maybe that prefix is a general solution
for other unit scales, but it seems like that's a job for whomever stewards
UDUNITS (just Unidata? Unidata+others? You would know better than me). I'm
no expert.

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Unidata/thredds/issues/454#issuecomment-191430846>.
>
Best wishes.


-- 
Sincerely,

Bob Simons
IT Specialist
Environmental Research Division
NOAA Southwest Fisheries Science Center
99 Pacific St., Suite 255A      (New!)
Monterey, CA 93940               (New!)
Phone: (831)333-9878            (New!)
Fax:   (831)648-8440
Email: bob.simons@noaa.gov

The contents of this message are mine personally and
do not necessarily reflect any position of the
Government or the National Oceanic and Atmospheric Administration.
<>< <>< <>< <>< <>< <>< <>< <>< <><
